sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/model/DataChangeRecord.java

```diff
@@ -208,6 +210,21 @@ public ChangeStreamRecordMetadata getMetadata() {
     return metadata;
   }
 
+  /**
+   * Estimates the size in bytes of a record when encoded using Avro.
+   *
+   * @return the number of bytes of the encoded object
+   */
+  public long bytesSize() {
```

Comment:
IIUC you're changing from Utf8 encode to avro encoding. 
Is this more expensive? It is per-element so it might matter should it be profiled?

Also I'm not sure AvroCoder is necessarily going to be used, it seems like it could depend on which coders are registered. Luke or someone else from Beam would be in a better spot to confirm best way to ensure same coder is used.

Reply:
Yes, performance is a worry I had, but during the execution it did not seem to have that much of an effect at the pipeline latency at scale. I will profile it though.

@lukecwik Would you be able to bring some light into the coders that are used by Beam here?

Reply:
Did some microbenchmarking (jmh) comparing `Ut8.encodedLength` and `AvroCoder.encode`. I did not commit the benchmarking class, as I think this is not an experiment that we will repeat often. See the results below:

  | Utf8Encoder (ms / op) | AvroCoder (ms / op)
-- | -- | --
p50 | 0.00 | 0.00
p90 | 0.00 | 0.00
p95 | 0.00 | 0.00
p99 | 0.00 | 0.01
p99.9 | 0.03 | 0.03
p99.99 | 0.39 | 0.09
p99.999 | 1.30 | 0.71
p99.9999 | 10.39 | 3.90
p100 | 22.74 | 16.73

`AvroCoder.encode` is fast / on par with the previous implementation.

Reply:
@nancyxu123 this benchmark (from the other comment)

Reply:
I would try to hide implementation specific methods like this within the place that actually needs them as we don't expect the end-user to need them.

Sam is right that you want to grab the coder when it is actually bound during PTransform#expand and use it during encoding/decoding. This would also align with moving this code to where it is needed.

Note that encoding each record is typically quite expensive and is why we estimate the PCollection size. Also coders specifically have a way to register a byte size observer instead of forcing the encode/decode round trip for coders that can give the size cheaply. See https://github.com/apache/beam/blob/48c70cc30742b45b17a1d18ece2f0d079bee3915/sdks/java/harness/src/main/java/org/apache/beam/fn/harness/data/PCollectionConsumerRegistry.java#L457 for additional details.